### PR TITLE
Improve basics aa1 to implement 1

### LINKED ISSUE #54: There should be an improvement to the react router structure
**Describe the feature or change you'd like**
Currently the structure in which the router is implemented does not seem neat. Therefore I suggest following changes:
- Move the router to `index.js`
- Export the App Component like `export withRouter(App)`to provide the history API

**Additional context**
Ref: https://developer.mozilla.org/en-US/docs/Web/API/History_API

### LINKED ISSUE #55: The SearchBar closes when clicking on the search label
**Describe the bug**
Currently the SearchBar closes when clicking on the search label.

**To Reproduce**
1. Go to 'snek.at'
2. Click on the SearchBar
3. Click on the Search label
4. See that the dropdown menu closes immediately

**Expected behavior**
I think the SearchBar shouldn't close?

**Additional context**
Ref: https://mdbootstrap.com/support/jquery/material-select-click-in-searchfield-closes-dropdown/ -> We use react and there it seems that it is not fixed yet by MDB

### LINKED ISSUE #57: There should be a template file for .env
**Describe the feature or change you'd like**
Currently there is no template file for the environment file. This sucks because I always has to ask my team about the content. @snek-at/frontend
 
It would be cool if there were a template file. If there is I could definitely sleep better.

### LINKED ISSUE #58: There should be regions in the App component to separate the methods
**Describe the feature or change you'd like**
There are a lot of methods inside the App component. Implementing some regions inside the component would be a good way.

**Additional context**
Ref: https://github.com/snek-at/front/blob/implement-1/src/App.js

### LINKED ISSUE #59: Upgrade react-scripts to v3.4.1
**Describe the feature or change you'd like**
Currently we are on v3.0.1 which produced 2 security vulnerabilities. Upgrading to v3.4.1 should fix it.

**Additional context**
Ref: https://npmjs.com/advisories/1426

### LINKED ISSUE #60: The SearchBar does not use the browser history API
**Describe the feature or change you'd like**
Currently switching to a other users is handled via `window.open(...)`. This shouldn't be the case and therefore replaced with the history API.

**Additional context**
Ref: https://developer.mozilla.org/en-US/docs/Web/API/History_API

### LINKED ISSUE #61: Some code uses platformData.user.username
**Describe the feature or change you'd like**
`platformData.user.username` is not valid in the current caching structure and should be replaced with `platformData.profile.username`.

### LINKED ISSUE #62: When switching the profile page while caching yours the cache gets fucked up
**Describe the feature or change you'd like**
Error behaviour:
1. Log in
2. See that caching is ongoing
3. Switch page via the SearchBar before caching is done
4. See error in console
5. Notice that your cache data might be fucked up

**Additional context**
This error occurs due to `intel.appendList()` is not only called for the logged user.
`snek-intel` does not use separated databases for each `intel.appendList()` iteration therefore the data seems to be inconsistent.

I would suggest to add some checks to prevent this in the `App.js`

### PR DESCRIPTION
**Pull request by schettn**
**Checklist:**

 
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [x] Have you updated or added documentation for the change?
 - [x] Have you tested your changes with successful results?

**Type of Changes:**

 
 - [x] Bug fix (non-breaking change which fixes an issue)
 - [x] New feature (non-breaking change which adds functionality)
 - [x] Documentation (non-breaking change which adds documentation)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change)

**What is the current behavior? (link to any open issues here)**
- Currently the router structure does not fit our needs. #54 
- Currently the SearchBar closes when clicking on the search label. #55 
- Currently `snek-intel:v0.1.0` is used which produces an error when generating talks. #56 
- Currently there is no `.env.template` file. #57 
- Currently there are no regions inside the App component. #58 
- Currently `react-scripts:v3.0.1` is used which has some security vulnerabilities. #59 
- Currently the SearchBar does not use the history API to switch profiles. #60 
- Currently some parts of the code uses the wrong notation for the username inside of platformData. #61 
- Currently it is not possible to switch a profile page while caching is ongoing. #62

**What is the new behavior (if this is a feature change)?**
- Now the router structure do fit our needs. Closes #54
- Now the SearchBar works fine. The label is removed now and therefore it is not possible to click on it.. It does not close :). Closes #55 
- Now `snek-intel:v0.1.1` is used. Therefore there are no more issues related to talk generation. #56 
- Now there is a  `.env.template` file. #57
- Now there are regions to separate methods in the App component. #58
- Now `react-scripts:v3.0.1` is used and the security vulnerabilities are fixed. #59 
- Now the SearchBar uses the history API. #60 
- Now the correct notation for the platformData username is used. #61 
- Now it is possible to switch the profile page while caching your own profile. #62

**Other information**:
- 🐍

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -13,10 +13,10 @@
     "react-chartjs-2": "^2.9.0",
     "react-dropzone": "^11.0.1",
     "react-router-dom": "^5.0.1",
-    "react-scripts": "3.0.1",
+    "react-scripts": "^3.4.1",
```
Comment (schettn) [inline package.json]: Move to devDependencies

File: src/App.js [hunk 2]
```diff
@@ -424,62 +461,62 @@ class App extends React.Component {
...
-      !this.state.caching &&
-      this.state.loggedUser?.username === platformData.user?.username
-    ) {
-      this.appendSourceObjects(fetchedUser.sources)
-        .then(async () => {
-          await this.intel.generateTalks(fetchedUser.sources);
-
-          let talks = await this.getAllTalks();
+    if (fetchedUser?.platformData) {
+      let platformData = fetchedUser.platformData;
```
Comment (schettn) [inline src/App.js:465]:
```suggestion
      let platformData = fetchedUser.platformData;
      
```

File: src/App.js [hunk 3]
```diff
@@ -424,62 +461,62 @@ class App extends React.Component {
...
+      if (
+        !this.state.caching &&
+        this.state.loggedUser?.username === platformData.profile?.username
+      ) {
+        this.appendSourceObjects(fetchedUser.sources)
+          .then(async () => {
+            await this.intel.generateTalks(fetchedUser.sources);
+            let talks = await this.getAllTalks();
```
Comment (schettn) [inline src/App.js]:
```suggestion
            await this.intel.generateTalks(fetchedUser.sources);
            
            let talks = await this.getAllTalks();
```

File: src/App.js [hunk 4]
```diff
@@ -424,62 +461,62 @@ class App extends React.Component {
...
+                }
               }
-            }
 
-            if (state) {
-              platformData.talks.push(talks[i]);
+              if (state) {
+                platformData.talks.push(talks[i]);
+              }
             }
```
Comment (schettn) [inline src/App.js:490]: Rework the indexes

**Review by schettn — COMMENTED** [review]
I will improve this asap

File: src/App.js [hunk 5]
```diff
@@ -424,62 +461,62 @@ class App extends React.Component {
...
-      this.state.loggedUser?.username === platformData.user?.username
-    ) {
-      this.appendSourceObjects(fetchedUser.sources)
-        .then(async () => {
-          await this.intel.generateTalks(fetchedUser.sources);
-
-          let talks = await this.getAllTalks();
+    if (fetchedUser?.platformData) {
+      let platformData = fetchedUser.platformData;
+      if (
```
Comment (pinterid) [inline src/App.js]:
```suggestion
      let platformData = fetchedUser.platformData;
      
      if (
```

File: src/App.js (hunk 3 above)
Comment (pinterid) [inline src/App.js]:
```suggestion
            await this.intel.generateTalks(fetchedUser.sources);
            
            let talks = await this.getAllTalks();
```

File: src/App.js [hunk 6]
```diff
@@ -544,53 +581,55 @@ class App extends React.Component {
...
+                updateCache: this.updateCache,
+                uploadTalk: this.uploadTalk,
+                deleteTalk: this.deleteTalk,
+                getTalk: this.getTalk,
+                login: this.login,
+                registerUser: this.registerUser,
+                fetchGitLabServers: this.fetchGitLabServers,
+                refetchRequired: this.refetchRequired,
+                usernameMatchesFetchedUsername: this
+                  .usernameMatchesFetchedUsername,
```
Comment (pinterid) [inline src/App.js:620]: Does this pass the max line length or why is this written in two lines?
```suggestion
                usernameMatchesFetchedUsername: this.usernameMatchesFetchedUsername,
```

Comment (schettn) [reply src/App.js:620]: Yep your prediction is completely right.

**Review by pinterid — CHANGES_REQUESTED** [review]
Looks good after some small changes.

**Review by pinterid — APPROVED** [review]
Looks good to merge.

**Comment by schettn** [comment]
I will remove @Aichnerc and @kleberbaum for time reasons.
Please have a sharp look at #52

**Pull request merged**